src/util/vecdeque.h

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_VECDEQUE_H
+#define BITCOIN_UTIL_VECDEQUE_H
+
+#include <util/check.h>
+
+#include <cstring>
+#include <memory>
+
+/** Data structure largely mimicking std::deque, but using single preallocated ring buffer.
+ *
+ * - More efficient and better memory locality than std::deque.
+ * - Most operations ({push_,pop_,emplace_,}{front,back}(), operator[], ...) are O(1),
+ *   unless reallocation is needed (in which case they are O(n)).
+ * - Supports reserve(), capacity(), shrink_to_fit() like vectors.
+ * - No iterator support.
+ * - Data is not stored in a single contiguous block, so no data().
+ */
+template<typename T>
+class VecDeque
+{
+    /** Pointer to allocated memory. Can contain constructed and uninitialized T objects. */
+    T* m_buffer{nullptr};
+    /** m_buffer + m_offset points to first object. m_offset < m_capacity. */
+    size_t m_offset{0};
+    /** Number of objects in the container. m_size < m_capacity. */
+    size_t m_size{0};
+    /** The size of m_buffer, expressed as a multiple of the size of T. */
+    size_t m_capacity{0};
+
+    size_t FirstPart() const noexcept { return std::min(m_capacity - m_offset, m_size); }
+
+    void Reallocate(size_t capacity)
+    {
+        Assume(capacity >= m_size);
+        Assume(m_capacity == 0 || m_offset < m_capacity);
```

Comment:
What is the purpose of this line here? It doesn't look like a pre-condition for the function arguments. It seems it should be used as a post-condition for `m_capacity` mutators (including this function) and `m_offset` mutators, no?